tests/optim/utils/test_atlas.py

```diff
@@ -0,0 +1,400 @@
+
+        xy_grid = atlas.normalize_grid(xy_grid, min_percentile=0.5)
+
+        xy_grid_expected = torch.tensor(
+            [
+                [0.0000, 0.0000],
+                [0.0000, 0.0000],
+                [0.0000, 0.0000],
+                [0.0000, 0.0000],
+                [0.0893, 0.0893],
+                [0.3169, 0.3169],
+                [0.5446, 0.5446],
+                [0.7723, 0.7723],
+                [1.0000, 1.0000],
+            ]
+        )
+
+        assertTensorAlmostEqual(self, xy_grid, xy_grid_expected, 0.001)
+
+    def test_normalize_grid_cuda(self) -> None:
+        if not torch.cuda.is_available():
+            raise unittest.SkipTest(
+                "Skipping normalize grid CUDA test due to not supporting CUDA."
+            )
+        if torch.__version__ < "1.7.0":
+            raise unittest.SkipTest(
+                "Skipping normalize grid CUDA test due to insufficient Torch version."
+            )
+        xy_grid = torch.arange(0, 2 * 3 * 3).view(3 * 3, 2).float().cuda()
+
+        xy_grid = atlas.normalize_grid(xy_grid)
+
+        xy_grid_expected = torch.tensor(
+            [
+                [0.0000, 0.0000],
+                [0.1250, 0.1250],
+                [0.2500, 0.2500],
+                [0.3750, 0.3750],
+                [0.5000, 0.5000],
+                [0.6250, 0.6250],
+                [0.7500, 0.7500],
+                [0.8750, 0.8750],
+                [1.0000, 1.0000],
+            ]
+        )
+
+        self.assertTrue(xy_grid.is_cuda)
+        assertTensorAlmostEqual(self, xy_grid.cpu(), xy_grid_expected)
+
+
+class TestCalcGridIndices(BaseTest):
+    def test_calc_grid_indices(self) -> None:
+        if torch.__version__ < "1.7.0":
+            raise unittest.SkipTest(
+                "Skipping calc grid indices test due to insufficient Torch version."
+            )
+        xy_grid = torch.arange(0, 2 * 3 * 3).view(3 * 3, 2).float()
+        xy_grid = atlas.normalize_grid(xy_grid)
+        indices = atlas.calc_grid_indices(xy_grid, grid_size=(2, 2))
+
+        expected_indices = [
+            [torch.tensor([0, 1, 2, 3, 4]), torch.tensor([4])],
+            [torch.tensor([4]), torch.tensor([4, 5, 6, 7, 8])],
+        ]
+
+        for list1, list2 in zip(indices, expected_indices):
+            for t1, t2 in zip(list1, list2):
+                assertTensorAlmostEqual(self, t1, t2)
+
+    def test_calc_grid_indices_extent(self) -> None:
+        if torch.__version__ < "1.7.0":
+            raise unittest.SkipTest(
+                "Skipping calc grid indices extent test due to insufficient Torch"
+                + " version."
+            )
+        xy_grid = torch.arange(0, 2 * 5 * 5).view(5 * 5, 2).float()
+        xy_grid = atlas.normalize_grid(xy_grid)
+        indices = atlas.calc_grid_indices(
+            xy_grid, grid_size=(1, 1), x_extent=(1.0, 2.0), y_extent=(1.0, 2.0)
+        )
+        assertTensorAlmostEqual(self, indices[0][0], torch.tensor([24]), 0)

```

Comment:
I see - 24 is the 24th (sample) index in the xy_grid (25 x 2) tensor, isn't it ?

Reply:
@NarineK 

It's the 25 item in the list (which counting from 0 is 24), because this test is setting the lower end of the ` x_extent and ``y_extent`. If we had the default extents, then the output would be:nar

```
[[tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,18, 19, 20, 21, 22, 23, 24])]]
```

Reply:
We start off with a tensor that looks like this:

```
tensor([[ 0.,  1.],
        [ 2.,  3.],
        [ 4.,  5.],
        [ 6.,  7.],
        [ 8.,  9.],
        [10., 11.],
        [12., 13.],
        [14., 15.],
        [16., 17.],
        [18., 19.],
        [20., 21.],
        [22., 23.],
        [24., 25.],
        [26., 27.],
        [28., 29.],
        [30., 31.],
        [32., 33.],
        [34., 35.],
        [36., 37.],
        [38., 39.],
        [40., 41.],
        [42., 43.],
        [44., 45.],
        [46., 47.],
        [48., 49.]])
```

Then we normalized the tensor so that it looks like this:

```
tensor([[0.0000, 0.0000],
        [0.0417, 0.0417],
        [0.0833, 0.0833],
        [0.1250, 0.1250],
        [0.1667, 0.1667],
        [0.2083, 0.2083],
        [0.2500, 0.2500],
        [0.2917, 0.2917],
        [0.3333, 0.3333],
        [0.3750, 0.3750],
        [0.4167, 0.4167],
        [0.4583, 0.4583],
        [0.5000, 0.5000],
        [0.5417, 0.5417],
        [0.5833, 0.5833],
        [0.6250, 0.6250],
        [0.6667, 0.6667],
        [0.7083, 0.7083],
        [0.7500, 0.7500],
        [0.7917, 0.7917],
        [0.8333, 0.8333],
        [0.8750, 0.8750],
        [0.9167, 0.9167],
        [0.9583, 0.9583],
        [1.0000, 1.0000]])
```

Then it's run through the `calc_grid_indices` function to group everything based on a grid that's x-axis is 1.0-2.0, and the y-axis is 1.0-2.0. Because we changed the starting part of the x & y axes, only the final 25th value with an index of 24 is used.